File: javascript/src/resources/responses.ts
import { addOutputText } from "../internal/output-text.js";
import { buildQuery } from "../internal/query.js";
import type { HTTPClient } from "../internal/http.js";
import { collectPage, Page } from "../pagination.js";
import { validateUniqueToolNames } from "../tool-validation.js";
import type { RequestOptions } from "../types/common.js";
import type {
  AgentResponse,
  CancelResponse,
  ListChildrenResponse,
  ListEventsParams,
  ListEventsResponse,
  ListResponsesParams,
  ListResponsesResponse,
  ResponseCreateParams,
  ResponseCreateParamsNonStreaming,
  ResponseCreateParamsStreaming,
  ResponseListItem,
} from "../types/responses.js";
import type { ResponseStreamEvent } from "../types/streaming.js";
import type { VolumeInfo } from "../types/volumes.js";

export class ResponsesResource {
  constructor(
    private readonly http: HTTPClient,
    private readonly path = "/v1/responses",
  ) {}

  create(params: ResponseCreateParamsNonStreaming, options?: RequestOptions): Promise<AgentResponse>;
  create(params: ResponseCreateParamsStreaming, options?: RequestOptions): Promise<AsyncIterable<ResponseStreamEvent>>;
  create(
    params: ResponseCreateParams,
    options?: RequestOptions,
  ): Promise<AgentResponse | AsyncIterable<ResponseStreamEvent>> {
    validateUniqueToolNames(params.tools);
    if (params.stream) {
      return this.http.stream<ResponseStreamEvent>(this.path, params, options);
    }
    return this.http.request<AgentResponse>("POST", this.path, params, options).then(addOutputText);
  }

  list(params: ListResponsesParams = {}, options?: RequestOptions): Promise<ListResponsesResponse> {
    return this.http.request<ListResponsesResponse>(
      "GET",
      `${this.path}${buildQuery({ limit: params.limit, page_token: params.page_token })}`,
      undefined,
      options,
    );
  }

  async listPage(params: ListResponsesParams = {}, options?: RequestOptions): Promise<Page<ResponseListItem, ListResponsesParams>> {
    return collectPage(
      (pageParams) => this.list(pageParams, options),
      params,
    );
  }

  listIterator(params: ListResponsesParams = {}, options?: RequestOptions): AsyncIterable<ResponseListItem> {
    const self = this;
    return {
      async *[Symbol.asyncIterator]() {
        const page = await self.listPage(params, options);
        yield* page.iterateAll();
      },
    };
  }

  retrieve(responseID: string, options?: RequestOptions): Promise<AgentResponse> {
    return this.http
      .request<AgentResponse>("GET", `${this.path}/${encodeURIComponent(responseID)}`, undefined, options)
      .then(addOutputText);
  }

  cancel(responseID: string, options?: RequestOptions): Promise<CancelResponse> {
    return this.http.request<CancelResponse>(
      "POST",
      `${this.path}/${encodeURIComponent(responseID)}/cancel`,
      undefined,
      options,
    );
  }

  listChildren(responseID: string, options?: RequestOptions): Promise<ListChildrenResponse> {
    return this.http.request<ListChildrenResponse>(
      "GET",
      `${this.path}/${encodeURIComponent(responseID)}/children`,
      undefined,
      options,
    );
  }

  listEvents(responseID: string, params: ListEventsParams = {}, options?: RequestOptions): Promise<ListEventsResponse> {
    return this.http.request<ListEventsResponse>(
      "GET",
      `${this.path}/${encodeURIComponent(responseID)}/events${buildQuery({
        after_sequence: params.after_sequence,
        view: params.view,
      })}`,
      undefined,
      options,
    );
  }

  retrieveVolume(responseID: string, options?: RequestOptions): Promise<VolumeInfo> {
    return this.http.request<VolumeInfo>(
      "GET",
      `${this.path}/${encodeURIComponent(responseID)}/volume`,
      undefined,
      options,
    );
  }
}

File: go/agentapi/responses.go
package agentapi

import (
	"context"
	"net/url"
)

type ResponsesService struct {
	http *httpClient
	path string
}

type ListResponsesParams struct {
	Limit            int
	PageToken        string
	SafetyIdentifier string
	UserID           string
}

type RetrieveResponseParams struct {
	SafetyIdentifier string
}

type ListEventsParams struct {
	AfterSequence int64
	View          string
}

func (s *ResponsesService) Create(ctx context.Context, params ResponseCreateParams, opts ...RequestOption) (*AgentResponse, error) {
	params.Stream = false
	if err := validateUniqueToolNames(params.Tools); err != nil {
		return nil, err
	}
	var out AgentResponse
	if err := s.http.requestJSON(ctx, "POST", s.path, params, &out, opts...); err != nil {
		return nil, err
	}
	addOutputText(&out)
	return &out, nil
}

func (s *ResponsesService) CreateStream(ctx context.Context, params ResponseCreateParams, opts ...RequestOption) (*ResponseStream, error) {
	params.Stream = true
	if err := validateUniqueToolNames(params.Tools); err != nil {
		return nil, err
	}
	resp, err := s.http.stream(ctx, "POST", s.path, params, opts...)
	if err != nil {
		return nil, err
	}
	return newResponseStream(resp), nil
}

func (s *ResponsesService) List(ctx context.Context, params ListResponsesParams, opts ...RequestOption) (*ListResponsesResponse, error) {
	var out ListResponsesResponse
	err := s.http.requestJSON(ctx, "GET", s.path+buildQuery(map[string]any{
		"limit":             params.Limit,
		"page_token":        params.PageToken,
		"safety_identifier": params.SafetyIdentifier,
		"user_id":           params.UserID,
	}), nil, &out, opts...)
	return &out, err
}

func (s *ResponsesService) Retrieve(ctx context.Context, responseID string, opts ...RequestOption) (*AgentResponse, error) {
	return s.RetrieveWithParams(ctx, responseID, RetrieveResponseParams{}, opts...)
}

func (s *ResponsesService) RetrieveWithParams(ctx context.Context, responseID string, params RetrieveResponseParams, opts ...RequestOption) (*AgentResponse, error) {
	var out AgentResponse
	err := s.http.requestJSON(ctx, "GET", s.path+"/"+url.PathEscape(responseID)+buildQuery(map[string]any{
		"safety_identifier": params.SafetyIdentifier,
	}), nil, &out, opts...)
	if err != nil {
		return nil, err
	}
	addOutputText(&out)
	return &out, nil
}

func (s *ResponsesService) Cancel(ctx context.Context, responseID string, opts ...RequestOption) (*CancelResponse, error) {
	var out CancelResponse
	err := s.http.requestJSON(ctx, "POST", s.path+"/"+url.PathEscape(responseID)+"/cancel", nil, &out, opts...)
	return &out, err
}

func (s *ResponsesService) ListChildren(ctx context.Context, responseID string, opts ...RequestOption) (*ListChildrenResponse, error) {
	var out ListChildrenResponse
	err := s.http.requestJSON(ctx, "GET", s.path+"/"+url.PathEscape(responseID)+"/children", nil, &out, opts...)
	return &out, err
}

func (s *ResponsesService) ListEvents(ctx context.Context, responseID string, params ListEventsParams, opts ...RequestOption) (*ListEventsResponse, error) {
	var out ListEventsResponse
	err := s.http.requestJSON(ctx, "GET", s.path+"/"+url.PathEscape(responseID)+"/events"+buildQuery(map[string]any{
		"after_sequence": params.AfterSequence,
		"view":           params.View,
	}), nil, &out, opts...)
	return &out, err
}

func (s *ResponsesService) RetrieveVolume(ctx context.Context, responseID string, opts ...RequestOption) (*VolumeInfo, error) {
	var out VolumeInfo
	err := s.http.requestJSON(ctx, "GET", s.path+"/"+url.PathEscape(responseID)+"/volume", nil, &out, opts...)
	return &out, err
}

func addOutputText(resp *AgentResponse) {
	if resp == nil || resp.OutputText != "" {
		return
	}
	var text string
	for _, item := range resp.Output {
		if item.Type != "message" {
			continue
		}
		for _, part := range item.Content {
			if part.Type == "output_text" {
				text += part.Text
			}
		}
	}
	resp.OutputText = text
}
